### DISCUSSION TITLE
Test Classloading Issue?

### DISCUSSION THREAD

**Discussion by The-Funk** [quarkusio/quarkus#43059] [Q&A]
I am having an issue in tests that I think is related to classloading.

I am using Panache Active Record pattern and have a few classes that are not CDI beans but use Panache Entity methods like findById. These include builders and factories, etc.

I have my entities loading in from a Quarkus extension which has a jandex index.

At runtime my entity methods work fine, however in tests, they sometimes will fail when called outside of the scope of something annotated with the Test annotation, and they are particularly fickle if there are any static methods involved in the test class that don't have a JUnit annotation.

Most of the time I can solve the issue by just adding ActivateRequestContext around the static method, and that seems to resolve things, but not always.

I have one case currently where I have a ParameterizedTest and even if I set the test lifecycle and make the argument producing method non-static I always get the same error:

```
java.lang.IllegalStateException: This method is normally automatically overridden in subclasses: did you forget to annotate your entity with @Entity?
        at io.quarkus.hibernate.orm.panache.common.runtime.AbstractJpaOperations.implementationInjectionMissing(AbstractJpaOperations.java:443)
```

I assume that the JUnit class loader is loading my entities before Quarkus, and so none of the augmentation that would normally be done by Quarkus is done on my entities. This makes it so that calling the entity method fails because no concrete implementation has been supplied yet.

I am using Quarkus 3.8.6

**Comment by The-Funk** [comment]
@holly-cummins help? 🙏

**Comment by holly-cummins** [comment]
I don't suppose you have a small reproducer, do you? :) I'm hopeful this will be fixed by https://github.com/orgs/quarkusio/projects/30/(the #34681 part), but in my first iteration it's not fixing *all* use cases related to parameterized tests.

**Reply by holly-cummins** [reply]
I also don't have any good suggestions for workarounds while #34681 is pending, sorry @The-Funk! Well, apart from "do not use parameterized tests," which is more "giving up" than "working around". :)

**Comment by The-Funk** [comment]
@holly-cummins I'll try to put together a minimal reproducer without the extension and with just a single Panache Entity. It might take a little doing but I think I can get it set up! Thank you for the info! I'm just glad that based on the info provided and your response, I feel like I have a grasp of what the issue is, so finding a solution should be possible.

I was looking at other suggestions of [using a delegate class](https://github.com/quarkusio/quarkus/issues/1724#issuecomment-534974823). I'm thinking if I can get the utility classes to be beans and then wrap them with a delegate defined in tests, that should resolve the issue. I'm also looking at making the utilities stateless and not have them talk to the database at all.
